src/member.rs

```diff
@@ -704,45 +748,43 @@ where
         }
     }
 
-    fn on_network_event(&mut self, event: NetworkEvent) {
-        match event {
-            NetworkEvent::MessageReceived(message, sender) => {
-                match ConsensusMessage::decode(&mut &message[..]) {
-                    Ok(message) => {
-                        self.on_consensus_message(message, sender);
-                    }
-                    Err(e) => {
-                        debug!(target: "network", "Error decoding message: {}", e);
-                    }
-                }
-            }
-        }
-    }
-
-    pub async fn run_session(
+    pub async fn run_session<N: Network + 'static>(
         mut self,
+        network: N,
         spawn_handle: impl SpawnHandle,
         exit: oneshot::Receiver<()>,
     ) {
         let (tx_consensus, consensus_stream) = mpsc::unbounded();
         let (consensus_sink, mut rx_consensus) = mpsc::unbounded();
         let (ordered_batch_tx, mut ordered_batch_rx) = mpsc::unbounded();
-        let (consensus_exit, exit_rx) = oneshot::channel();
+        let (consensus_exit, exit_stream) = oneshot::channel();
         let config = self.config.clone();
         let sh = spawn_handle.clone();
         debug!(target: "rush-member", "Spawning party for a session with config {:?}", self.config);
-        spawn_handle.spawn("consensus/root", async move {
+        spawn_handle.spawn("member/consensus", async move {
             consensus::run(
                 config,
                 consensus_stream,
                 consensus_sink,
                 ordered_batch_tx,
                 sh,
-                exit_rx,
+                exit_stream,
             )
             .await
         });
         self.tx_consensus = Some(tx_consensus);
+        let (alert_sink, mut incoming_alerts) = mpsc::unbounded();
+        let (outgoing_alerts, alert_stream) = mpsc::unbounded();
+        let (unit_sink, mut incoming_units) = mpsc::unbounded();
+        let (outgoing_units, unit_stream) = mpsc::unbounded();
+        let (network_exit, exit_stream) = oneshot::channel();
+        spawn_handle.spawn("member/network", async move {
+            NetworkHub::new(network, unit_stream, unit_sink, alert_stream, alert_sink)
```

Comment:
Wouldn't it be clearer to construct NetworkHub in the constructor of Member, store it as Option, and take it here? It would also allow to store `self.outgoing_[units,alerts]` directly and get rid of unwraps in corresponding send methods.

Reply:
The unwraps won't happen after the full refactor is complete – a property of the general design I'm trying to work towards is that no one creates their own channels. `Member` would create the `outgoing_[units,alerts]` channels, but then pass them on to a component that would actually take care of unit-related messages (the parts of `Member` that currently do that would be moved to that component). In the end `Member` should just start all the other components and itself stop existing as a struct (the way `consensus` now just exists as `run`).

In general having `Option` fields in structs that mean "this is here always, but we cannot initialize it during construction" is always bad design, so I definitely did not want to leave it like that forever. I assumed it would be as temporary as `self.tx_consensus` which already follows this anti-pattern.

Reply:
Ok, so I wait for the final version :)

Reply:
Oh, I don't want to complete the full refactor in this PR, that would be a bit much. <_<"

Reply:
sure sure. It took me some time to fix network hub analog in aleph-node to not contain Options, so I know that it's best done on final version. Waiting for the next PRs then :)